src/bsoncxx/builder/bson/bson.hpp

```diff
@@ -0,0 +1,75 @@
+// Copyright 2020 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <bsoncxx/config/prelude.hpp>
+#include <bsoncxx/types/bson_value/value.hpp>
+
+namespace bsoncxx {
+BSONCXX_INLINE_NAMESPACE_BEGIN
+namespace builder {
+using namespace bsoncxx::types;
+
+class bson {
+   public:
+    template <typename T>
+    bson(T value) : _value{value} {}
```

Comment:
In short, yes, it's used in all of the tests.

This class has two constructors: one that takes an initializer list of BSON builder values and one that attempts to implicitly convert any value of type T to a BSON builder value (above). Take this example:
```
bson doc {"foo", {"bar", 1}};    
```
Since this _is_ an initializer list but _not_ of BSON builder values, the compiler must check whether each value in the list can be implicitly converted to a BSON builder value. So the constructor above would be called on `"foo"`, `"bar"` and `1`. Then we'll have:
 ```
bson doc{bson("foo"), {bson("bar"), bson(1)}}
```
The nested document is still not a BSON builder value. It is an initializer list of BSON builder values. Thus, the initializer list constructor is called, and those two values and combined into one document. Afterward, we're left with:
 ```
 /* the nested document is now a BSON builder value whose underlying type is a doc */
bson doc{bson("foo"), bson({"bar", 1})}
```
Now we finally have an initializer list of BSON builder values. The initializer list constructor is called to combine these two BSON builder values into one document, just like for the nested one.

Reply:
Ah, that is actually fascinating.  Thank you for explaining!

Reply:
Np